File: src/leetcode/deletenodeinabst/DeleteNodeInABST.java
package leetcode.deletenodeinabst;

import trees.TreeNode;

public class DeleteNodeInABST {

    public TreeNode deleteNode(TreeNode root, int key) {
        return delete(root, null, key);
    }

    private TreeNode delete(TreeNode root,
                            TreeNode parent,
                            int key) {
        if (root != null) {
            if (root.val == key) {
                if (root.left == null) {
                    return root.right;
                } else if (root.right == null) {
                    return root.left;
                } else {
                    // both children are not null so
                    // find the min of the right subtree
                    int minFromRight = findMin(root.right);
                    root.val = minFromRight;
                    root.right = deleteNode(root.right, minFromRight);

                    return root;
                }
            } else {
                if (root.val > key) {
                    root.left = delete(root.left, root, key);
                } else {
                    root.right = delete(root.right, root, key);
                }
                return root;
            }
        } else {
            return null;
        }
    }

    private int findMin(TreeNode root) {
        int min = Integer.MAX_VALUE;
        while (root != null) {
            min = Math.min(min, root.val);
            root = root.left;
        }

        return min;
    }

}

File: python-code/delete_node_in_a_bst.py
from typing import *
from tree_node import TreeNode


class Solution:
    '''
    - Find the node of the existing tree which has val == key

    '''

    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:

        if not root:
            return None

        if root.val == key:
            if not root.right and not root.left:
                return None

            elif not root.left:
                return root.right
            elif not root.right:
                return root.left
            else:
                # both left and right are populated
                # Find the successor of the node to be deleted: The successor of the node
                # to be deleted is the smallest node of the right subtree
                successor_value = self._find_minimum(root.right)
                root.val = successor_value
                root.right = self.deleteNode(root.right, successor_value)
        elif key > root.val:
            root.right = self.deleteNode(root.right, key)
        else:
            root.left = self.deleteNode(root.left, key)

        return root

    def _find_minimum(self, root: Optional[TreeNode]) -> int:
        min_val = float('inf')
        while root:
            min_val = min(min_val, root.val)
            root = root.left

        return min_val
